From 3ad3a1fa41d3973b21b6c01525b24c7b9f489022 Mon Sep 17 00:00:00 2001
From: Luc Maisonobe <luc@apache.org>
Date: Fri, 11 Nov 2011 23:39:34 +0000
Subject: [PATCH] Fixed a bad interaction between step handlers and event
 handlers in ODE integrators.

The problem was due to an internal array from the step interpolator that
was changed by the step handlers. These array was expected to contain
the state at event time, but was overridden later on. the solution is
simply to clone the array. Note that the fact the array is reused *is*
documented in the javadoc and is a feature of the interpolators ...

JIRA: MATH-706

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/trunk@1201105 13f79535-47bb-0310-9956-ffa450edef68
---
 .../java/org/apache/commons/math/ode/AbstractIntegrator.java  | 2 +-
 src/site/xdoc/changes.xml                                     | 4 ++++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/math/ode/AbstractIntegrator.java b/src/main/java/org/apache/commons/math/ode/AbstractIntegrator.java
index 8f315e3349..e09a92206d 100644
--- a/src/main/java/org/apache/commons/math/ode/AbstractIntegrator.java
+++ b/src/main/java/org/apache/commons/math/ode/AbstractIntegrator.java
@@ -318,7 +318,7 @@ public int compare(EventState es0, EventState es1) {
 
                 // trigger the event
                 interpolator.setInterpolatedTime(eventT);
-                final double[] eventY = interpolator.getInterpolatedState();
+                final double[] eventY = interpolator.getInterpolatedState().clone();
                 currentEvent.stepAccepted(eventT, eventY);
                 isLastStep = currentEvent.stop();
 
diff --git a/src/site/xdoc/changes.xml b/src/site/xdoc/changes.xml
index b54cdaffa9..a12d6d1a1f 100644
--- a/src/site/xdoc/changes.xml
+++ b/src/site/xdoc/changes.xml
@@ -52,6 +52,10 @@ The <action> type attribute can be add,update,fix,remove.
     If the output is not quite correct, check for invisible trailing spaces!
      -->
     <release version="3.0" date="TBD" description="TBD">
+      <action dev="luc" type="fix" due-to="MATH-706" >
+        Fixed a bad interaction between step handlers and event handlers in
+        ODE integrators.
+      </action>
       <action dev="luc" type="add" due-to="Jan Kotek" >
         Added array constructor and getter for Vector2D and Vector3D.
       </action>
